### DISCUSSION TITLE
Provide a way to access the extended custom colors in JS/TS

### DISCUSSION THREAD

**Discussion by IonelLupu** [tailwindlabs/tailwindcss#11895] [Ideas]
We have this tailwind.config.js file:
```js
module.exports = {
    content: ['./pages/**/*.{js,ts,jsx,tsx}', './app/**/*.{js,ts,jsx,tsx}', './components/**/*.{js,ts,jsx,tsx}'],
    theme: {
        extend: {
            colors: {
                primary: '#013729',
                forest: {
                    50: '#E8F2EC',
                    200: '#B9D9C5',
                    700: '#15803D',
                    800: '#166534',
                    950: '#013729'
                }
            },
        },
    }
}
```
We are using AntDesign and it's config sits in a .tsx file like this:
```tsx
import colors from 'tailwindcss/colors'

// ... some code

<ConfigProvider
            theme={{
                token: {
                    colorWarning: colors.orange['600'],
                    colorWarningBorder: colors.orange['200'],
                    colorWarningBg: colors.orange['50'],
                    colorError: colors.red['600'],
                    colorErrorBorder: colors.red['200'],
                    colorErrorBg: colors.red['50'],


                    colorSuccess: colors.forest['700'], // Cannot read properties of undefined (reading '700')
                    colorSuccessBg: colors.forest['50'],
                    colorSuccessBorder: colors.forest['200'],
                },
           }}
/>
```

As you can see, we are able to use Tailwind's built-in colors for "error" and "warning", but we can't use the custom "forest" color for "success" even if we added it in the config file.

Using the CSS class directly works. But we want to change the Ant Design theme to use the same colors too.

**Comment by tlgreg** [comment]
You can load your whole resolved config with `tailwindcss/resolveConfig`. [Relevant docs](https://tailwindcss.com/docs/configuration#referencing-in-java-script)

**Reply by IonelLupu** [reply]
Ok. It work after compiling the code, but the IDE is not pleased with that 😅 . Is there a way to fix this?

![image](https://github.com/tailwindlabs/tailwindcss/assets/4083652/46deeeb7-f07b-474f-b1b5-532513bfbcb7)

It doesn't even like it's own properties:
![image](https://github.com/tailwindlabs/tailwindcss/assets/4083652/11d89eca-6529-426b-aeb7-3caba9b67e15)

**Reply by tlgreg** [reply]
This looks like a type issue, how the types are done right now you always get a generic config type.
For colors this means it's "infinitely" recursive, it can't know when it is a leaf property.

*I myself didn't work with `resolveConfig` since tailwind has its own types.* The workaround I can see without deeper exploration is to export your colors separately from the config and use those directly. Or even just the types and assert those, like this would still provide you the default resolved config too:

```ts
// tailwind.config.ts
import type { Config } from 'tailwindcss'
import defaultColors from 'tailwindcss/colors'

export type DefaultColors = typeof defaultColors

const colors = {
  primary: '#013729',
  forest: {
    50: '#E8F2EC',
    200: '#B9D9C5',
    700: '#15803D',
    800: '#166534',
    950: '#013729',
  },
} as const

export type Colors = typeof colors

const config: Config = {
  // ...
  theme: {
    extend: {
      colors,
    },
  },
}

export default config
```

```tsx
import resolveConfig from 'tailwindcss/resolveConfig'
import tailwindConfig, { type Colors, type DefaultColors } from '../tailwind.config'

const tw = resolveConfig(tailwindConfig)
const { theme } = tw as unknown as {
  theme: (typeof tw)['theme'] & { colors: DefaultColors & Colors }
}

// ...
```

![Code_-_Insiders_BFDcI1UBke](https://github.com/tailwindlabs/tailwindcss/assets/666791/a90ba7c6-52e4-43c5-97b9-7b71f51a35a1)

---

A real solution down the line would need to be a different `resolveConfig` type that is less generic in its result, i.e. treat the output of config even more differently than the input. Output ideally should be just an object literal for both `resolveConfig` return type and the `defaultConfig`, `defaultTheme` files.
